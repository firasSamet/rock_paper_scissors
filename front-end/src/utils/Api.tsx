import axios from "axios";

const BASE_API = "http://localhost:8080/api";

const getSessionId = async () => await axios.post(`${BASE_API}/session`);

const resetSession = async (id: string) =>
  await axios.delete(`${BASE_API}/session/${id}`);

const playRound = async (id: string) =>
  await axios.post(`${BASE_API}/games/${id}`);

const getRounds = async (id: string) =>
  await axios.get(`${BASE_API}/games/${id}`);

const getStats = async () => await axios.get(`${BASE_API}/session`);

export const api = {
  getSessionId,
  resetSession,
  playRound,
  getRounds,
  getStats,
};
