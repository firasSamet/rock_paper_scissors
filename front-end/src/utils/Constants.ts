export enum ButtonTypes {
    "PRIMARY" = "primary",
    "SECONDARY" = "secondary",
  }
  
export enum Choice {
    ROCK = "ROCK",
    PAPER = "PAPER",
    SCISSORS = "SCISSORS",
  }

export enum GameResult {
    "PLAYER1" = "Player 1 won",
    "PLAYER2" = "Player 2 won",
    "DRAW" = "Draw",
}
  