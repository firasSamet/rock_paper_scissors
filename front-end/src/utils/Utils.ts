export const parseResult = (result: number) =>
  ["Draw", "Player 1 won", "Player 2 won"].at(result);