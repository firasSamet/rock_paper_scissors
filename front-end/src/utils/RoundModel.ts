import { Choice } from "./Constants";

export interface RoundModel {
    firstPlayerChoice: Choice
    secondPlayerChoice: Choice
    gameResult: number
}