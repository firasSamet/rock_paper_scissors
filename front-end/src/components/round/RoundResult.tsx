import React from "react";
import classnames from "classnames";

import { PlayerChoice } from "../playerChoice/PlayerChoice";

import { RoundModel } from "../../utils/RoundModel";

import "./Round.scss";
import { parseResult } from "../../utils/Utils";

interface RoundResultProps {
  roundResults: RoundModel;
}

export const RoundResult = ({ roundResults }: RoundResultProps) => {
  const gameResultClassNames = classnames("game-result", {
    draw: !roundResults?.gameResult,
  });

  return (
    <div className="round-wrapper">
      <div className="round-container">
        <PlayerChoice choice={roundResults?.firstPlayerChoice} player={1} />
        <PlayerChoice choice={roundResults?.secondPlayerChoice} player={2} />
      </div>
      {!!roundResults?.firstPlayerChoice && (
        <h1 className={gameResultClassNames}>
          {parseResult(roundResults.gameResult)}
        </h1>
      )}
    </div>
  );
};
