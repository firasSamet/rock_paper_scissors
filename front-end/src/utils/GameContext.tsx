import React, { createContext, useContext, useEffect, useState } from "react";
import { RoundModel } from "./RoundModel";

export interface GameState {
  sessionId: string;
  gameData: RoundModel[];
  gameStats: number[];
}

interface GameContext {
  gamestate: GameState;
  updateGameContext: (newValue: GameState) => void;
}

const GameContext = createContext<GameContext | undefined>(undefined);

export function useGameContext(): GameContext {
  const context = useContext(GameContext);
  if (!context) {
    throw new Error("useGameContext must be used within a GameContextProvider");
  }
  return context;
}

export function GameContextProvider({
  children,
}: {
  children: React.ReactNode;
}) {
  const [gamestate, setGamestate] = useState<GameState>(() => {
    const storedState = localStorage.getItem("gamestate");
    return storedState
      ? JSON.parse(storedState)
      : { sessionId: "", gameData: [] };
  });

  useEffect(() => {
    localStorage.setItem("gamestate", JSON.stringify(gamestate));
  }, [gamestate]);

  const updateGameContext = (newValue: GameState) => {
    setGamestate(newValue);
  };

  return (
    <GameContext.Provider value={{ gamestate, updateGameContext }}>
      {children}
    </GameContext.Provider>
  );
}
