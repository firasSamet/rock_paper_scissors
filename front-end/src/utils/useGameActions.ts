import { useGameContext } from './GameContext';

import {api} from './Api';

const useGameActions = () => {
  const { gamestate, updateGameContext } = useGameContext();

  const getSessionId = async () => {
    if (!gamestate?.sessionId) {
      try {
        const response = await api.getSessionId();
        const sessionId = response.data.sessionID;
        updateGameContext({ ...gamestate, sessionId });
      } catch (error) {
        console.error('Error retrieving session ID:', error);
      }
    }
  };

  const resetSession = async () => {
    try {
      await api.resetSession(gamestate.sessionId);
      updateGameContext({ ...gamestate, gameData: [] });
    } catch (error) {
      console.error('Error resetting session:', error);
    }
  };

  const playRound = async () => {
    try {
      const response = await api.playRound(gamestate.sessionId);
      const rounds = response.data;
      updateGameContext({ ...gamestate, gameData: rounds });
      return rounds.at(-1);
    } catch (error) {
      console.error('Error playing round:', error);
    }
  };

  const getRounds = async () => {
    try {
      const response = await api.getRounds(gamestate.sessionId);
      updateGameContext({ ...gamestate, gameData: response.data });
    } catch (error) {
      console.error('Error retrieving rounds:', error);
    }
  };

  return { getSessionId, resetSession, playRound, getRounds };
};

export default useGameActions;
