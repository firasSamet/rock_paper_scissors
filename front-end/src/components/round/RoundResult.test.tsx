import React from "react";
import { render, screen } from "@testing-library/react";
import { Choice } from "../../utils/Constants";
import { RoundResult } from "./RoundResult";

describe("RoundResult component", () => {
  test("Should render correctly", () => {
    const wrapper = render(
      <RoundResult
        roundResults={{
          firstPlayerChoice: Choice.PAPER,
          secondPlayerChoice: Choice.ROCK,
          gameResult: 1,
        }}
      />
    );

    expect(wrapper.baseElement).toMatchSnapshot();
  });

  test("should display correct winner", async () => {
    render(
      <RoundResult
        roundResults={{
          firstPlayerChoice: Choice.PAPER,
          secondPlayerChoice: Choice.ROCK,
          gameResult: 1,
        }}
      />
    );
    const winner = await screen.findByText(/PLAYER 1 WON/i);

    expect(winner.classList.contains("draw")).toBe(false);
  });

  test("should display a draw game correctly", async () => {
    render(
      <RoundResult
        roundResults={{
          firstPlayerChoice: Choice.ROCK,
          secondPlayerChoice: Choice.ROCK,
          gameResult: 0,
        }}
      />
    );
    const winner = await screen.findByText(/DRAW/i);

    expect(winner.classList.contains("draw")).toBe(true);
  });

  test("should not display a result if there is no round data", () => {
    render(<RoundResult />);
    const winner = screen.queryByText(/DRAW/i);

    expect(winner).toBeFalsy();
  });
});
