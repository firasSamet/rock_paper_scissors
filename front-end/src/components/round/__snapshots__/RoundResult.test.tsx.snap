// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RoundResult component Should render correctly 1`] = `
<body>
  <div>
    <div
      class="round-wrapper"
    >
      <div
        class="round-container"
      >
        <div
          class="player-choice-wrapper"
        >
          <div
            class="player-choice-container"
          >
            <div>
              <h3>
                Player 
                1
              </h3>
              <p>
                PAPER
              </p>
            </div>
          </div>
        </div>
        <div
          class="player-choice-wrapper"
        >
          <div
            class="player-choice-container"
          >
            <div>
              <h3>
                Player 
                2
              </h3>
              <p>
                ROCK
              </p>
            </div>
          </div>
        </div>
      </div>
      <h1
        class="game-result"
      >
        Player 1 won
      </h1>
    </div>
  </div>
</body>
`;
